fenix/app/src/main/java/org/mozilla/fenix/translations/TranslationsDialogStore.kt

```diff
@@ -0,0 +1,225 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+package org.mozilla.fenix.translations
+
+import mozilla.components.concept.engine.translate.Language
+import mozilla.components.concept.engine.translate.TranslationError
+import mozilla.components.lib.state.Action
+import mozilla.components.lib.state.Middleware
+import mozilla.components.lib.state.State
+import mozilla.components.lib.state.Store
+
+/**
+ * The [TranslationsDialogStore] holds the [TranslationsDialogState] (state tree).
+ *
+ * The only way to change the [TranslationsDialogState] inside
+ * [TranslationsDialogStore] is to dispatch an [Action] on it.
+ */
+class TranslationsDialogStore(
+    initialState: TranslationsDialogState,
+    middlewares: List<Middleware<TranslationsDialogState, TranslationsDialogAction>> = emptyList(),
+) : Store<TranslationsDialogState, TranslationsDialogAction>(
+    initialState,
+    ::translationsDialogStateReducer,
+    middlewares,
+) {
+    init {
+        dispatch(TranslationsDialogAction.ExpectedTranslations)
+    }
+}
+
+/**
+ * The current state of the Translations bottom sheet dialog.
+ *
+ * @property isTranslated The page is currently translated.
+ * @property isInProgress The page is currently attempting a translation.
+ * @property positiveButtonType Can be enabled,disabled or in progress.
+ * @property error An error that can occur during the translation process.
+ * @property dismissDialogState Whether the dialog bottom sheet should be dismissed.
+ * @property initialFrom Initial "from" language, based on the translation state and page state.
+ * @property initialTo Initial "to" language, based on the translation state and page state.
+ * @property fromLanguages Translation menu items to be shown in the translate from dropdown.
+ * @property toLanguages Translation menu items to be shown in the translate to dropdown.
+ * @property translatedPageTitle Title of the bottom sheet dialogue if the page was translated.
+ */
+data class TranslationsDialogState(
+    var isTranslated: Boolean = false,
+    val isInProgress: Boolean = false,
+    val positiveButtonType: PositiveButtonType? = null,
+    val error: TranslationError? = null,
+    val dismissDialogState: DismissDialogState? = null,
+    val initialFrom: Language? = null,
+    val initialTo: Language? = null,
+    val fromLanguages: List<Language>? = null,
+    val toLanguages: List<Language>? = null,
+    val translatedPageTitle: String? = null,
+) : State
```

Comment:
As mentioned in my [comment](https://github.com/mozilla-mobile/firefox-android/pull/5242#discussion_r1482166568), I think we can observe most of these properties directly from the `BrowserStore` and leave this state for Fenix-specific items.